### DISCUSSION TITLE
Scala, How to run a Slick database action inside an action?

### DISCUSSION THREAD

**Discussion by mikael-jarvinen** [slick/slick#2355] [Questions] [answered]
I have a many to many relationship between Category and OfferRequest models, with CategoryOfferRequest model in between containing relationships between both these models.

And I am trying to implement a function to fetch Categories by OfferRequest.
My problem is in the Case with SimpleRelation("categoriesByOfferRequest"). First I have to query the OfferRequest ids from CategoryOfferRequest table and then query Category table with those ids

My code 
```
def getCategoriesByRelations(rel: RelationIds[Category]) = db.run(
  Categories.filter(category =>
    rel.rawIds.collect({
      case (SimpleRelation("categoryByCategory"), ids: Seq[Int]) => 
        category.upperCategoryId inSet ids
      case (SimpleRelation("categoriesByOfferRequest"), ids: Seq[Int]) =>
        category.id inSet //CategoryId here
    }).foldLeft(true: Rep[Boolean])(_ || _)
  ).result
)
```

What I would want to have is in place of "//CategoryId here" an action to fetch categoryIds from CategoryOfferRequests table, like so
```
CategoryOfferRequests.filter(_.offerRequestId inSet ids).result
```

I am new to slick, and don't yet know alot about actions, but is there a way to do this, or some other way?

**Comment by nafg** [comment, marked as answer]
Hi, you don't have to compose Actions, you can compose Queries.

You should be able to write

```scala
category.id.in(CategoryOfferRequests.map(_.offerRequestId))
```

When you realize that Scala's `map` is SQL `select` you can see how that becomes `id IN (SELECT offer_request_id FROM category_offer_request)`


To learn more about Slick, in addition to https://scala-slick.org there's a great free book you can find online called Essential Slick.
